backend/utils/pubsub.js

```diff
@@ -0,0 +1,254 @@
+  return JSON.stringify({ type, data })
+}
+
+/**
+ * Creates a pubsub server instance.
+ *
+ * @param {(http.Server|https.Server)} server - A Node.js HTTP/S server to attach to.
+ * @param {Object?} options
+ * {object?} clientHandlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {object?} serverHandlers - Custom handlers for server events.
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = options.clientHandlers || {}
+  server.customServerHandlers = options.serverHandlers || {}
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers[name]
+        const defaultHandler = defaultServerHandlers[name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = {
+  /**
+   * Emitted when a connection handshake completes.
+   *
+   * @param socket - The client socket that connected.
+   * @param request - The http GET request sent by the socket.
+   */
+  connection (socket, request) {
```

Comment:
Should these parameters have their type specified? (Either here in the arguments, or in the comment above?)

Reply:
This can be done to improve the Flowtype coverage, but is not required to pass the check according to the current .flowconfig.

Reply:
Right, I'm less concerned about Flowtype coverage, and more interested in simply understanding what these parameters are as I don't have this Websocket API memorized. Maybe in the comment you could include a link to where one could read up about these parameters? Or just a name or something to make it easier to look up the reference to these?

Reply:
However to keep it simple it's probably best to annotate them with just `Object` for now. Which is unfortunate for the type checking accuracy, but we don't have a Flowtype definition file for the `ws` library, and the exact type of the request argument is something like `http$IncomingMessage<net$Socket | tls$TLSSocket>`

Reply:
Yeah, like I said, the flowtype annotation is less important to me than maybe a link to some place where more can be read up about these parameters so that people who are unfamiliar with the API have some place to go to if they're interested in knowing what these are.